### PR TITLE
[Doc] Add descriptions for C++ Client API docs

### PR DESCRIPTION
**Pull request by Anonymitaet**
Fix a part of https://github.com/apache/pulsar/issues/9282 and add descriptions for the following `.h` files.

- Authentication.h
- Client.h
- Consumer.h
- CryptoKeyReader.h
- EncryptionKeyInfo.h
- KeySharedPolicy.h
- Logger.h
- Message.h
- MessageBuilder.h
- MessageId.h
- MessageRoutingPolicy.h
- Reader.h
- TopicMetadata.h

### CODE REVIEW

**Review by sijie — APPROVED** [review]

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 1]
```diff
@@ -157,9 +300,37 @@ class PULSAR_PUBLIC AuthAthenz : public Authentication {
...
     AuthAthenz(AuthenticationDataPtr&);
     ~AuthAthenz();
+
+    /**
+     * Create an AuthAthenz with a ParamMap
+     *
+     * The required parameter keys are “tenantDomain”, “tenantService”, “providerDomain”, “privateKey”, and
+     * “ztsUrl”
+     *
+     * @param parameter the key-value to construct ZTS client
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Authentication.h:310]:
```suggestion
     * @param params the key-value to construct ZTS client
```

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 2]
```diff
@@ -178,14 +349,54 @@ class Oauth2TokenResult {
     Oauth2TokenResult();
     ~Oauth2TokenResult();
 
+    /**
+     * Set the access token string
+     *
+     * @param AccessToken the access token string
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Authentication.h:355]:
```suggestion
     * @param accessToken the access token string
```

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 3]
```diff
@@ -178,14 +349,54 @@ class Oauth2TokenResult {
...
+     */
     Oauth2TokenResult& setAccessToken(const std::string& accessToken);
+
+    /**
+     * Set the ID token
+     *
+     * @param IdToken the ID token
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Authentication.h:362]:
```suggestion
     * @param idToken the ID token
```

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 4]
```diff
@@ -178,14 +349,54 @@ class Oauth2TokenResult {
...
+     */
     Oauth2TokenResult& setIdToken(const std::string& idToken);
+
+    /**
+     * Set the refresh token which can be used to obtain new access tokens using the same authorization grant
+     * or null for none
+     *
+     * @param RefreshToken the refresh token
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Authentication.h:370]:
```suggestion
     * @param refreshToken the refresh token
```

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 5]
```diff
@@ -178,14 +349,54 @@ class Oauth2TokenResult {
...
+     */
     Oauth2TokenResult& setRefreshToken(const std::string& refreshToken);
+
+    /**
+     * Set the token lifetime
+     *
+     * @param ExpiresIn the token lifetime
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Authentication.h:377]:
```suggestion
     * @param expiresIn the token lifetime
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 6]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
+     */
     void createProducerAsync(const std::string& topic, ProducerConfiguration conf,
                              CreateProducerCallback callback);
 
+    /**
+     * Subscribe to a given topic and subscription combination with the default ConsumerConfiguration
+     *
+     * @param topic the topic name
+     * @param subscription the subscription name
+     * @return ResultOk if it subscribes to the topic successfully
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:113]:
```suggestion
     * @param[out] consumer the consumer instance to be returned
     * @return ResultOk if it subscribes to the topic successfully
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 7]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
+     */
     Result subscribe(const std::string& topic, const std::string& subscriptionName, Consumer& consumer);
+
+    /**
+     * Subscribe to a given topic and subscription combination with the customized ConsumerConfiguration
+     *
+     * @param topic the topic name
+     * @param subscription the subscription name
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:121]:
```suggestion
     * @param subscriptionName the subscription name
     * @param conf the consumer configuration object
     * @param[out] consumer the consumer instance to be returned
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 8]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
+     */
     Result subscribe(const std::string& topic, const std::string& subscriptionName,
                      const ConsumerConfiguration& conf, Consumer& consumer);
 
+    /**
+     * Asynchronously subscribe to a given topic and subscription combination with the default
+     * ConsumerConfiguration
+     *
+     * @param topic the topic name
+     * @param subscription the subscription name
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:132]:
```suggestion
     * @param subscriptionName the subscription name
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 9]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
     void subscribeAsync(const std::string& topic, const std::string& subscriptionName,
                         SubscribeCallback callback);
+
+    /**
+     * Asynchronously subscribe to a given topic and subscription combination with the customized
+     * ConsumerConfiguration
+     *
+     * @param topic the topic name
+     * @param subscription the subscription name
+     * @param configuration the customized ConsumerConfiguration
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:145]:
```suggestion
     * @param subscriptionName the subscription name
     * @param conf the customized ConsumerConfiguration
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 10]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
+     * @param callback the callback that is triggered when a given topic and subscription combination with the
+     * customized ConsumerConfiguration are asynchronously subscribed successfully or not
+     */
     void subscribeAsync(const std::string& topic, const std::string& subscriptionName,
                         const ConsumerConfiguration& conf, SubscribeCallback callback);
 
     /**
-     * subscribe for multiple topics under the same namespace.
+     * Subscribe to multiple topics under the same namespace.
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:153]:
```suggestion
     * Subscribe to multiple topics under the same namespace.
     *
     * @param topics a list of topic names to subscribe to
     * @param subscriptionName the subscription name
     * @param[out] consumer the consumer instance to be returned
```

File: pulsar-client-cpp/include/pulsar/Client.h [hunk 11]
```diff
@@ -86,42 +86,140 @@ class PULSAR_PUBLIC Client {
...
      */
     Result subscribe(const std::vector<std::string>& topics, const std::string& subscriptionName,
                      Consumer& consumer);
+
+    /**
+     * Subscribe to multiple topics with the customized ConsumerConfiguration under the same namespace
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Client.h:159]:
```suggestion
     * Subscribe to multiple topics with the customized ConsumerConfiguration under the same namespace
     *
     * @param topics a list of topic names to subscribe to
     * @param subscriptionName the subscription name
     * @param conf the customized ConsumerConfiguration
     * @param[out] consumer the consumer instance to be returned
```

**Review by merlimat — APPROVED** [review]
👍

**Comment by Anonymitaet** [comment]
@merlimat many thanks for your comments, I've incorporated them.

**Review by Jennifer88huang — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Hi @BewareMyPower the tests were re-run many times these 2 days but still failed, do you have any clues? Thanks!

**Comment by Anonymitaet** [comment]
@yunze thanks for your help, all tests passed.

**Pull request merged**